pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java

```diff
@@ -39,62 +38,63 @@
+      if (scalarFunction.enabled()) {
+        if (!scalarFunction.name().isEmpty()) {
+          FunctionRegistry.registerFunction(scalarFunction.name(), method);
+        } else {
+          FunctionRegistry.registerFunction(method);
+        }
+      }
+    }
+    LOGGER.info("Initialized FunctionRegistry within {}ms", System.currentTimeMillis() - startTimeMs);
+    _initialized = true;
   }
 
   /**
-   * Given a function name and a set of argument types, asserts that a corresponding function
-   * was registered during construction and returns it
+   * Registers a method with the name of the method.
    */
-  public static FunctionInfo getFunctionByNameWithApplicableArgumentTypes(String functionName,
-      Class<?>[] argumentTypes) {
-    FunctionInfo functionInfo = getFunctionByName(functionName);
-    Preconditions.checkArgument(functionInfo.isApplicable(argumentTypes));
-    return functionInfo;
-  }
-
-  public static void registerFunction(Method method) {
+  public static synchronized void registerFunction(Method method) {
     registerFunction(method.getName().toLowerCase(), method);
   }
 
-  public static void registerFunction(String name, Method method) {
+  /**
+   * Registers a method with the given function name.
+   */
+  public static synchronized void registerFunction(String functionName, Method method) {
     FunctionInfo functionInfo = new FunctionInfo(method, method.getDeclaringClass());
-    _functionInfoMap.put(name, functionInfo);
+    FUNCTION_INFO_MAP.put(functionName, functionInfo);
   }
 
-  public static boolean containsFunctionByName(String funcName) {
-    return _functionInfoMap.containsKey(funcName.toLowerCase());
-  }
-
-  static {
-    try {
-
-      Reflections reflections = new Reflections(
-          new ConfigurationBuilder().setUrls(ClasspathHelper.forPackage("org.apache.pinot"))
-              .setScanners(new MethodAnnotationsScanner()));
-
-      Set<Method> methodSet = reflections.getMethodsAnnotatedWith(ScalarFunction.class);
-      for (Method method : methodSet) {
-        ScalarFunction scalarFunction = method.getAnnotation(ScalarFunction.class);
-        if (scalarFunction.enabled()) {
-          if (!scalarFunction.name().isEmpty()) {
-            FunctionRegistry.registerFunction(scalarFunction.name(), method);
-          } else {
-            FunctionRegistry.registerFunction(method);
-          }
-        }
-      }
-
-    } catch (Exception e) {
-      LOGGER.error("Caught exception when registering function", e);
-      throw new IllegalStateException(e);
-    }
+  /**
+   * Returns the {@link FunctionInfo} associated with the given function name, or {@code null} if there is no method
+   * registered under the name. This method should be called after the FunctionRegistry is initialized and all methods
+   * are already registered.
+   */
+  @Nullable
+  public static FunctionInfo getFunctionByName(String functionName) {
+    return FUNCTION_INFO_MAP.get(functionName.toLowerCase());
```

Comment:
Could you make a function like `canonicalize(functionName)` and use it here and also in `registerFunction` method where FUNCTION_INFO_MAP gets populated so in future we won't miss converting to lowercase?

Reply:
Added